### ISSUE TITLE
Discussion: child elements cannot determine whether they are "active" or not

### ISSUE THREAD

**Issue by caridy** [WICG/webcomponents#817]
In the process to optimize a web component, we often encounter a situation where child elements from the light-dom are unable to determine, reliable, if they are slotted or not into its parent shadow. Abstract example:

```html
    <x-foo>
        <x-bar>
            <x-baz></x-baz>
        </x-bar>
    </x-foo>
```

Depending on the state of `<x-foo>`, it attaches the `<slot>` element or not into its shadow, causing `<x-bar>` to be slotted accordingly.

And whether or not that happens, is not observable in `<x-bar>` and `<x-baz>` without doing some gymnastics that are not trivial. Considering that `<x-bar>` and `<x-baz>` could be very complex, and heavy, e.g. (content of a tab, or content for a panel, etc), it will be interesting to explore a way for those elements to get a notification when they become slotted (part of the flat tree). At the moment, it seems that checking for `this.assignedSlot` is good enough, but unfortunately it is a pull process, and if you get slotted after insertion into the light-dom, you have no way to tell as far as I can tell. 

This will be more self-evident if we ever introduce manual slotting.

**Comment by rniwa** [comment]
I don't think detecting whether an element is slotted or not is sufficient since a parent node might be the one getting slotted as well.

In the past, there was a discussion about adding a callback to when a node gets "attached" in the sense of render tree in WebKit sense. That is when an element is visible (does not have `display: none`) part of a document.

**Comment by caridy** [comment]
Yes, `attachedCallback` and `detachedCallback` will definitely solve this. I'm curious about those previous discussions, any link to notes or some other material about that? Or at least, can you give us the gist of the conversation, any particular point of contention about such new hooks?

**Comment by emilio** [comment]
I'd be concerned with exposing such callbacks because different engines have different kinds of behavior here. For example Gecko can end up rebuilding ancestor layout trees, if it turns out that dynamically inserting something is hard, like with some IB-split cases and table cases that WebKit and Blink get wrong. A naive way of defining this would result in a "detached" and "attached" callback for no reason.

A callback for getting into the flat tree and out, or such, seems more sensible to me, or alternatively, maybe, an explicit API that gives you whether you're rendered? There's other APIs that already expose that (`getClientRects()` returns an empty list if you don't have a layout box for example).

**Comment by caridy** [comment]
@emilio about the `getClientRects()`, that's a pull process... how do you know when to call it? That's the biggest challenge!

**Comment by emilio** [comment]
Sure, I see that, just wanted to note that even though `assignedSlot` is not enough to detect whether an element is rendered (as Ryosuke noted) we're already exposing other ways to detect it.

If you want a callback instead, callbacks when getting into / outside of the flat tree may be the easiest thing to define.

**Comment by caridy** [comment]
another interesting suggestion that I got from a co-worker who is working on perf, is that the [Page Visibility API](https://www.w3.org/TR/2017/PR-page-visibility-2-20171017/) can be probably extended to cover ShadowRoot as well. Something like:

```js
class Foo extends HTMLElement {
    constructor() {
        ...
    }
    connectedCallback() {
        const videoElement = this.#shadowRoot.getElementById("videoElement");
        // Autoplay the video if element is visible
        if (this.#shadowRoot.visibilityState == "visible") {
            videoElement.play();
        }

        // Handle shadowRoot visibility change events
        this.#shadowRoot.addEventListener('visibilitychange', () => {
            if (this.#shadowRoot.visibilityState == "hidden") {
                videoElement.pause();
            } else {
                videoElement.play();
            }
        }, false);
    }
}
```

**Comment by caridy** [comment]
@rniwa @domenic @annevk any opinion on this? This happens to be one of the points mentioned by Rich Harris in his post. We need to have a reliable way to avoid carrying on a lot of operations for fragments of the DOM that are not slotted. Until now, `Element.isConnected` was sufficient (although not observable), but with the introduction of the ShadowDOM, this isn't a reliable way anymore. As for custom elements, connectedCallback (which could be considered analogous to `isConnected`) is not sufficient.

**Comment by justinfagnani** [comment]
@caridy I think Rich's criticism is more about laziness than visibility or connectedness. In many "over the top" frameworks you can give a component some children, and that component can decide whether or not to even create the DOM for those children. With web components, if a component has children, they're necessarily already instantiated.

I think the solution there is to recognize that `<template>` is the lazy operator for HTML. A component can receive template children and clone them on demand.

**Comment by caridy** [comment]
@justinfagnani as I mentioned above, I believe it depends on the component, and the content of its shadow (and how defensive that component can be when it comes to visibility), and it our analysis, we have determined that by using detection at the first level of non-slotted/visible elements is good enough, because we haven't seen components with huge templates, it is normally a composition of other complex components.

As for scoped/lazy templates, that introduces a brand new set a problems, because the owner of the template doesn't have access or control over the instances of that template instantiated by the receiver of the template. It is just a brand new class a problems to be solved. What we are proposing is just a very small signal that can take us very far, orthogonal to the concept of scoped/lazy templates.

**Comment by justinfagnani** [comment]
I'm merely talking about the relation to Rich's criticism, which is about laziness. Yes, the template owner doesn't have control over the elements, but this is exactly like over-the-top frameworks where a component can modify its children before rendering.

**Comment by annevk** [comment]
Pointer to Rich's post? If you know that an element can be slotted, but you don't know whether the parent element is initialized and has created its shadow tree yet, does `this.parentNode.matches(":defined")` not give you the information you need?

**Comment by caridy** [comment]
@annevk the original post is here: https://dev.to/richharris/why-i-don-t-use-web-components-2cia (specifically, 5. Composition section). The example he uses is the following:

```html
<p>Toggle the section for more info:</p>
<toggled-section>
  <html-include src="./more-info.html"/>
</toggled-section>
```

Whether `html-include` component is defined or not is irrelevant, what is relevant is that with shadow DOM, that component must be instantiated (while other frameworks don't do that), but I don't think that's something we can address at this point. In our analysis (after looking at few thousands components at salesforce), it seems that if `html-include` is able to detect and observe its state in terms of slotting, to determine whether or not it is visible/slotted, it could decide whether or not to populate its shadow. By cutting the DOM tree branches shorts, we can avoid doing unnecessary job, that seems enough. Imagine that the component will instantiate conditionally dozens of components as part of its shadow root.
